src/controller/CHIPDeviceController.cpp

```diff
@@ -212,7 +212,10 @@ CHIP_ERROR DeviceController::InitControllerNOCChain(const ControllerInitParams &
     ReturnErrorOnFailure(ConvertX509CertToChipCert(params.controllerNOC, nocSpan));
     ReturnErrorOnFailure(ExtractNodeIdFabricIdFromOpCert(nocSpan, &nodeId, &fabricId));
 
-    auto * fabricTable            = params.systemState->Fabrics();
+    auto * systemState = params.systemState ? params.systemState : mSystemState;
```

Comment:
So this is trying to handle the case when someone called Init() without key material and is not calling InitControllerNOCChain with a different set of params that they did not bother to set the SystemState on?

It's not obvious why we would want to support that, nor is it clear to me why we would ever want to support the SystemState on these params being different from mSystemState in practice... Except that on the Init() codepath we currently set mSystemState after calling this method.  I guess so we can do early returns before retaining the system state...

Concretely: should we not error out of both params.systemState and mSystemState are non-null and they are different?  That seems like it would just not work very well.

Reply:
> So this is trying to handle the case when someone called Init() without key material and is not calling InitControllerNOCChain with a different set of params that they did not bother to set the SystemState on?

More or less. The key problem here is not that "they did not bother" to set the SystemState parameter, there is just no public API to get the same system state and pass it to `InitControllerNOCChain`

> Concretely: should we not error out of both params.systemState and mSystemState are non-null and they are different? That seems like it would just not work very well.

In fact I can't see the scenario where they are different. But such error will break current behavior. While this PR preserves backwards compatibility

Reply:
> there is just no public API to get the same system state and pass it to InitControllerNOCChain

The system state lives on the controller factory, and can be gotten from there as needed.

> But such error will break current behavior.

That current behavior seems very broken already.

Reply:
Well, InitControllerNOCChain should be taking a const systemstate, since it does not in fact modify it, right?

Reply:
Yes, but `InitControllerNOCChain` takes the same `ControllerInitParams` struct as `Init`, and Init calls `systemState->Retain()`

Reply:
Ah, I see, we are passing the system state in a weird way here, because we took a shortcut...

Ideally we would have `InitControllerNOCChain` take a type that correctly represents the const-ness, or not, of the things involved....  Because what we have right now is a giant footgun in all sorts of ways.

Reply:
Well. Looks like its very weird scenario to call `Init` with one `systemState` and `InitControllerNOCChain` with the other. I'd suggest to remove systemState from `InitControllerNOCChain`'s params and always use `mSystemState`. I can update current PR

Reply:
I would be totally fine with that....

Reply:
@bzbarsky-apple I've updated PR. Sorry for long delay - I've forget about it.

Reply:
And... I've reverted it back.
Because I've realized that `Init` calls `InitControllerNOCChain` before it stores `mSystemState`, so it can safely fail on `InitControllerNOCChain` fail.
So in fact there are two scenarios of `InitControllerNOCChain` call:
- inside the `Init` function before the `mSystemState` is installed. In this case `systemState` should be passed via params
- after the `Init` call. In such case we can use `mSystemState`

So probably fallback is the optimal for now

Reply:
Wouldn't it be better to just change Init() to store mSystemState, then unset it on failure @j0tunn?  That seems much better than having this confusing "we will look for the state in various places, guess where we will find it" behavior.